frontend/src/data/parseReportData.ts

```diff
@@ -174,54 +174,73 @@ export const implementationFromData = (
-  loaderData: Record<string, ReportData>,
+/**
+ * Prepare a summarized implementation report for the
+ * passed implementation using all the dialect reports data
+ * that was fetched.
+ */
+export const prepareImplementationReport = (
+  allReportsData: Map<Dialect, ReportData>,
+  langImplementation: string,
 ) => {
-  const allImplementations: Record<string, ImplementationDialectCompliance> =
-    {};
-  const dialectCompliance: Record<string, Record<string, Partial<Totals>>> = {};
+  let implementationReport: ImplementationReport | null = null;
+  const dialectCompliance = new Map<Dialect, Partial<Totals>>();
+
+  // FIXME: This magic prefix is duplicated from the backend side,
+  //        and probably we can separate handling this in Implementation
+  //        class (when we have it).
+  const image = `ghcr.io/bowtie-json-schema/${langImplementation}`;
 
   for (const [
     dialect,
     { implementationsResults, runMetadata },
-  ] of Object.entries(loaderData)) {
-    dialectCompliance[dialect] = calculateImplementationTotal(
-      implementationsResults,
-    );
+  ] of allReportsData.entries()) {
+    const implementationResults = implementationsResults.get(image);
+
+    if (implementationResults) {
+      updateDialectCompliance(
+        dialectCompliance,
+        dialect,
+        implementationResults,
+      );
+
+      const implementation = runMetadata.implementations.get(image)!;
 
-    for (const [id, implementation] of runMetadata.implementations.entries()) {
-      if (!allImplementations[id]) {
-        allImplementations[id] = {
+      if (!implementationReport) {
+        implementationReport = {
           implementation,
-          dialectCompliance: {},
+          dialectCompliance,
+        };
+      } else {
+        implementationReport = {
+          implementation: Object.assign(
+            {},
+            implementationReport.implementation,
+            implementation,
+          ),
+          dialectCompliance: new Map([
+            ...implementationReport.dialectCompliance,
+            ...dialectCompliance,
+          ]),
         };
-      }
-      if (dialectCompliance[dialect][id]) {
-        allImplementations[id].dialectCompliance[dialect] =
-          dialectCompliance[dialect][id];
       }
     }
   }
 
-  return allImplementations;
+  return implementationReport;
 };
 
-const calculateImplementationTotal = (
-  implementationsResults: Map<string, ImplementationResults>,
-) => {
-  const implementationResult: Record<string, Partial<Totals>> = {};
-
-  Array.from(implementationsResults.entries()).forEach(([key, value]) => {
-    implementationResult[key] = {
-      erroredTests: value.erroredTests,
-      skippedTests: value.skippedTests,
-      failedTests: value.failedTests,
-    };
+const updateDialectCompliance = (
+  dialectCompliance: Map<Dialect, Partial<Totals>>,
+  dialect: Dialect,
+  implementationResults: ImplementationResults,
+) =>
+  dialectCompliance.set(dialect, {
```

Comment:
This is just a one line function that's only used in one place isn't it? Any reason not to just inline this above?

Reply:
Cool I will do it inline. Just a sec

Reply:
@Julian doing inline now 👍